android/runtime/v8/src/native/JSException.cpp

```diff
@@ -12,41 +15,55 @@
 
 #include "JSException.h"
 
+#define MAX_STACK 10
+
 using namespace v8;
 
 namespace titanium {
 
 Local<Value> JSException::fromJavaException(v8::Isolate* isolate, jthrowable javaException)
 {
-	JNIEnv *env = JNIScope::getEnv();
+	JNIEnv* env = JNIScope::getEnv();
 	if (!env) {
 		return GetJNIEnvironmentError(isolate);
 	}
 
-	env->ExceptionDescribe();
-
 	bool deleteRef = false;
 	if (!javaException) {
 		javaException = env->ExceptionOccurred();
-		env->ExceptionClear();
 		deleteRef = true;
 	}
+	env->ExceptionClear();
 
-	//env->ExceptionDescribe();
-
-	jstring message = (jstring) env->CallObjectMethod(javaException, JNIUtil::throwableGetMessageMethod);
-	if (!message) {
+	jstring javaMessage = (jstring) env->CallObjectMethod(javaException, JNIUtil::throwableGetMessageMethod);
+	if (!javaMessage) {
 		return THROW(isolate, "Java Exception occurred");
 	}
+	const char* messagePtr = env->GetStringUTFChars(javaMessage, NULL);
+	std::stringstream message;
+	message << messagePtr;
 
-	Local<Value> jsMessage = TypeConverter::javaStringToJsString(isolate, env, message);
-	env->DeleteLocalRef(message);
+	jobjectArray frames = (jobjectArray) env->CallObjectMethod(javaException, JNIUtil::throwableGetStackTraceMethod);
+	jsize frames_length = env->GetArrayLength(frames);
+	for (int i = 0; i < (frames_length > MAX_STACK ? MAX_STACK : frames_length); i++) {
+		jobject frame = env->GetObjectArrayElement(frames, i);
+		jstring javaStack = (jstring) env->CallObjectMethod(frame, JNIUtil::stackTraceElementToStringMethod);
+
+		const char* stackPtr = env->GetStringUTFChars(javaStack, NULL);
+		message << std::endl << "    " << stackPtr;
+
+		env->ReleaseStringUTFChars(javaStack, stackPtr);
+		env->DeleteLocalRef(javaStack);
+	}
+	message << std::endl;
 
+	env->ReleaseStringUTFChars(javaMessage, messagePtr);
```

Comment:
any reason we don't just do this cleanup right away after appending the messagePtr to the stream? (i.e. line 45 or so)